### ISSUE TITLE
dockerCall and dockerCheckOutput broken

### ISSUE THREAD

**Issue by glennhickey** [BD2KGenomics/toil#1959]
I realize they've apparently been deprecated but the shims in place do not work.  Get errors like 

    workDir=work_dir, outfile = outfile)
  File "/home/hickey/dev/toil-vg/deptest/lib/python2.7/site-packages/toil/lib/docker.py", line 55, in dockerCall
    return subprocessDockerCall(job=job, *args, **kwargs)
TypeError: subprocessDockerCall() got multiple values for keyword argument 'job'

because both job and *args get passed through resulting in a duplicate.

**Comment by glennhickey** [comment]
even running subprocessDockerCall directly doesn't work when I pass an outfile

  File "/home/hickey/dev/toil-vg/deptest/lib/python2.7/site-packages/toil/lib/docker.py", line 177, in subprocessDockerCall
    out = subprocess.check_output(call, **params)
  File "/usr/lib/python2.7/subprocess.py", line 566, in check_output
    raise ValueError('stdout argument not allowed, it will be overridden.')
ValueError: stdout argument not allowed, it will be overridden.

**Comment by DailyDreaming** [comment]
That doesn't look like an error with dockerCall.  Can I see the command you are running?

**Comment by glennhickey** [comment]
This is the code that broke when upgrading to toil 3.12.0 from 3.11.0
https://github.com/vgteam/toil-vg/blob/master/src/toil_vg/vg_common.py#L143-L191

The line is:
ret = dockerCall(job, tool, parameters=parameters, dockerParameters=docker_parameters,
                             workDir=work_dir, outfile = outfile)

You can see in docker.py that dockerCall is passing both job and *args (which contains job) down to subprocessDockerCall

https://github.com/BD2KGenomics/toil/blob/master/src/toil/lib/docker.py#L55

Even working around this, I get the second error (ValueError: stdout argument not allowed, it will be overridden).  Because when I specify an outfile this happens

https://github.com/BD2KGenomics/toil/blob/master/src/toil/lib/docker.py#L172-L173

but subprcess.check_output doesn't take stdout

https://github.com/BD2KGenomics/toil/blob/master/src/toil/lib/docker.py#L177

**Comment by DailyDreaming** [comment]
Just made a PR: https://github.com/BD2KGenomics/toil/pull/1960

I can modify it to switch between check_output and check_call but it seems like by calling stdout explicitly you're trying to get check_output functionality with check_call.

**Comment by glennhickey** [comment]
I am not specifying stdout explicitly

**Comment by DailyDreaming** [comment]
Alright, I've modified the PR to include the same output once more.  Please let me know how it works.  I would encourage you to move to the api for the future however.

**Comment by glennhickey** [comment]
How do I redirect stdout to a file with the new API (equivalent of outfile parameter in old dockerCall)?

**Comment by DailyDreaming** [comment]
If you call the function using `stdout=True`, say `output = apiDockerCall(job, cmd, stdout=True)`, then `output` should be the stdout string, which can be redirected to a file.  It's not explicitly built-in however, and I could add it if you like.

**Comment by glennhickey** [comment]
I run many commands that dump hundreds of gigs to stdout.  I can't store these in strings, and would need an option to pipe directly into a file to use this api.

**Comment by DailyDreaming** [comment]
Looking into this further, I believe it's just setting the `log_config` variable.  As long as this is set, specifying `stdout=True` and/or `stderr` should write stdout or stderr to the log file (which seems to replicate the old CLI functionality).

http://docker-py.readthedocs.io/en/stable/containers.html

**Comment by glennhickey** [comment]
That's good.  I will test this now.  Can you please tell me how to specify the log file?

**Comment by DailyDreaming** [comment]
I haven't used it yet, but it looks like this should work:

`{"log-driver": "json-file",
  "log-opts": {"labels": "any-label-you-might-want-to-add"}}`

Other log-opts:

max-size: "100k" (prevent the log file from getting too large)
max-file: "20" (prevent too many files from being generated from multiple)
env: "os,ubuntu" (write env variable into log file)

**Comment by glennhickey** [comment]
I don't mean to be obtuse, I'm still not understanding.  I need to do something like this:
     
     with open("/data/huge_binary_file.bin", "w") as f:
          # SUBSTITUTE:  dockerCall(image, params, etc, outfile=f)
          apiDockerCall(image, params, etc, log_config=XXX) # <--- WITH THIS

I'm not sure how Docker's JSON log driver is doing this.

**Comment by DailyDreaming** [comment]
It should be something like this:

    apiDockerCall(job,
                  image='quay.io/ucsc_cgl/samtools-latest',
                  working_dir=working_dir,
                  containerName='container--l2i3cm5o9-name',
                  detach=True,
                  remove=True,
                  {"log-driver": "json-file", "log-opts": {"labels": "any-label-you-might-want-to-add"}},
                  stdout=True
    )

**Comment by DailyDreaming** [comment]
It pipes stdout to a log file, which should be found here: /var/lib/docker/containers/<container id>/<container id>-json.log

**Comment by glennhickey** [comment]
Still not seeing it (my file "f" appears nowhere in your command).  Will assume that piping to file is not supported and stick with the old API.  This is a fairly major regression, at least from toil-vg's perspective.

**Comment by DailyDreaming** [comment]
That's fair.  I'd asked a few people earlier if there were additional features needed and none came up so it wasn't a concern when I wrote (mostly I went by the unittests).  I suspect I may need to spend a bit more time with your code to see what you're trying to do exactly.  I'll see if I can update this and test later.

I've made the review changes to address your earlier comment.  If you approve we'll merge and hopefully subprocessDockerCall should run smoothly until I can see how to replicate your use case.

**Comment by glennhickey** [comment]
OK, I guess we have two choices.  

Either Toil will support the subprocess versions of Docker call, in which case you'd be best off just putting back the old code (and tests), which worked fine, exactly as they were.

Or Toil will not support the subprocess version of Docker call.   In which case, I'll just move the old code int o toil-vg.     

I'm honeslty fine either way.  What I don't want to do is spend time debugging the current implementation of subprocess docker call which has been broken for no apparent reason, and clearly isn't under any testing (as it doesn't even run in https://github.com/BD2KGenomics/toil/pull/1960).

**Comment by DailyDreaming** [comment]
The intention was not to break the old code, and I went to some effort to ask people to test it.  Apologies for removing subprocess's ability to use the 'stdout'=filename feature, as that in particular was never a part of the unittests, and I overlooked it.  I missed that the subprocessDockerCall tests were not running, and that was not intended either.  The PR should be fixed, and as Joel mentioned, it looks like it was a single variable omitted that was preventing it from running.  Sorry for the headache.
